### PR TITLE
OGC Api features suport for non 4326 layers

### LINKED ISSUE #50051: QGIS OAPIF client to support OGC Api features part 2 - CRS
### Feature description

QGIS should be able to query OAPIF servers with user specified CRS when supported by the server. 

Now the query doesn't include crs request, and is always wgs84, even when explicitly specified by the user in QGIS datasource manager:

image

request:
https://avoin-paikkatieto.maanmittauslaitos.fi/maastotiedot/features/v1/collections/rakennus/items?limit=1000&bbox=22.24913618031373375,60.4686065088286,22.28079736364862029,60.47853041708438582' 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by domi4484**
Read layer CRS from collections request instead of hard coded

fixes #50051

### CODE REVIEW

**Review by 3nids — APPROVED** [review]

File: src/providers/wfs/oapif/qgsoapifprovider.cpp [hunk 1]
```diff
@@ -702,27 +714,32 @@ void QgsOapifFeatureDownloaderImpl::run( bool serializeFeatures, long long maxFe
...
+
+    if ( ! rect.isNull() )
     {
       url += ( hasQueryParam ? QStringLiteral( "&" ) : QStringLiteral( "?" ) );
       url += QStringLiteral( "bbox=%1,%2,%3,%4" )
-             .arg( qgsDoubleToString( minx ),
-                   qgsDoubleToString( miny ),
-                   qgsDoubleToString( maxx ),
-                   qgsDoubleToString( maxy ) );
+             .arg( qgsDoubleToString( rect.xMinimum() ),
```
Comment (rouault) [inline src/providers/wfs/oapif/qgsoapifprovider.cpp:739]: if the CRS of the layer is not the default OGC:CRS84 one, then the bbox-crs parameter should be specified (cf https://docs.opengeospatial.org/is/18-058r1/18-058r1.html#_parameter_bbox_crs), and care should be taken to do the proper axis order inversion for CRS with latitude-longitude or northing-easting order. Cf requirement 23.B of https://docs.opengeospatial.org/is/17-069r4/17-069r4.html#_parameter_bbox that mentions in a rather cryptic way that the first value should be "Lower left corner, coordinate axis 1". Slide 9 of https://files.justobjects.nl/presentation/geonovum-2021-api-strategie/OGC-API-Features.pdf makes it more explicit: "bbox-crs: URI of the CRS in which the coordinates are in the bbox parameter; axis order follows the axis order in the CRS"

**Review by rouault — CHANGES_REQUESTED** [review]

File: src/providers/wfs/oapif/qgsoapifcollection.cpp [hunk 2]
```diff
@@ -297,6 +298,19 @@ bool QgsOapifCollection::deserialize( const json &j )
...
 
+  if ( j.contains( "crs" ) )
+  {
+    const auto crsUrls = j["crs"];
+    if ( crsUrls.is_array() )
+    {
+      for ( const auto &crsUrl : crsUrls )
+      {
+        QString crs = QString::fromStdString( crsUrl.get<std::string>() );
+        mLayerMetadata.setCrs( QgsCoordinateReferenceSystem::fromOgcWmsCrs( crs ) );
```
[the remark this answers is not in the document]
Comment (domi4484) [reply src/providers/wfs/oapif/qgsoapifcollection.cpp:309]: No, in fact I think the first would make more sense, until support for multiple is implemented.

File: src/providers/wfs/oapif/qgsoapifprovider.cpp (hunk 1 above)
Comment (domi4484) [reply src/providers/wfs/oapif/qgsoapifprovider.cpp:739]: Is there a way to get the CRS in the format with the url? Or could I add a function in `QgsCoordinateReferenceSystem` to return a string in the format like `http://www.opengis.net/def/crs/OGC/1.3/CRS84`? A problem is that the version is not parsed...

About the axis order, from what I could see the currentRect is in the layer CRS, respecting also axis order inverted or not. In that case the clamping would be wrong for crs with inverted axis. I'll check that

Comment (elpaso) [reply src/providers/wfs/oapif/qgsoapifprovider.cpp:739]: @domi4484 have a look to QString QgsServerApiUtils::crsToOgcUri

Comment (elpaso) [reply src/providers/wfs/oapif/qgsoapifprovider.cpp:739]: If that suits your needs you might consider moving it from server to core.

Comment (domi4484) [reply src/providers/wfs/oapif/qgsoapifprovider.cpp:739]: Thanks @elpaso , I moved it to QgsCoordinateReferenceSystem and declared the one in QgsServerApiUtils deprecated.

@rouault bbox-crs paramter in request is implemented in the last commit

File: src/core/proj/qgscoordinatereferencesystem.cpp [hunk 3]
```diff
@@ -1495,6 +1495,29 @@ QgsRectangle QgsCoordinateReferenceSystem::bounds() const
   return rect;
 }
 
+QString QgsCoordinateReferenceSystem::toOgcUri() const
+{
+  const auto parts { authid().split( ':' ) };
+  if ( parts.length() == 2 )
+  {
+    if ( parts[0] == QLatin1String( "EPSG" ) )
+      return  QStringLiteral( "http://www.opengis.net/def/crs/EPSG/9.6.2/%1" ).arg( parts[1] ) ;
```
Comment (rouault) [inline src/core/proj/qgscoordinatereferencesystem.cpp:1504]: the usual practice is to use the dummy 0 generic version:
```suggestion
      return  QStringLiteral( "http://www.opengis.net/def/crs/EPSG/0/%1" ).arg( parts[1] ) ;
```

http://www.opengis.net/def/crs/EPSG/0/4326 is a dereferencable URL, whereas http://www.opengis.net/def/crs/EPSG/9.6.2/4326 is not

File: src/providers/wfs/oapif/qgsoapifprovider.cpp (hunk 1 above)
Comment (rouault) [reply src/providers/wfs/oapif/qgsoapifprovider.cpp:739]: > bbox-crs paramter in request is implemented in the last commit

axis inversion should also be added if QgsCoordinateReferenceSystem::hasAxisInverted()  ( to be checked how it behaves with OGC:CRS84, where axis order is longitude, latitude, contrary to EPSG:4326 which is latitude, longitude)

File: tests/src/python/test_provider_oapif.py [hunk 4]
```diff
@@ -326,7 +334,7 @@ def testBbox(self):
                  "geometry": {"type": "Point", "coordinates": [-65.32, 78.3]}}
             ]
         }
-        with open(sanitize(endpoint, '/collections/mycollection/items?limit=1000&' + ACCEPT_ITEMS), 'wb') as f:
+        with open(sanitize(endpoint, '/collections/mycollection/items?limit=1000&bbox=-180,-90,180,90&bbox-crs=http://www.opengis.net/def/crs/EPSG/9.6.2/4326&' + ACCEPT_ITEMS), 'wb') as f:
```
Comment (rouault) [inline tests/src/python/test_provider_oapif.py:337]: for EPSG:4326, that should be bbox=-90,-180,90,180

File: src/core/proj/qgscoordinatereferencesystem.cpp (hunk 3 above)
Comment (domi4484) [reply src/core/proj/qgscoordinatereferencesystem.cpp:1504]: Ok thanks, and I suppose for the OGC url the same?

Comment (rouault) [reply src/core/proj/qgscoordinatereferencesystem.cpp:1504]: > and I suppose for the OGC url the same?

hum, I'm more used in seeing 1.3 being used... But both http://www.opengis.net/def/crs/OGC/0/CRS84 and http://www.opengis.net/def/crs/OGC/1.3/CRS84 work

File: src/providers/wfs/oapif/qgsoapifprovider.cpp (hunk 1 above)
Comment (domi4484) [reply src/providers/wfs/oapif/qgsoapifprovider.cpp:742]: Ok now it should be good. At the beginning I thought erroneously that the received rectangle was already right inverted.

**Comment by domi4484** [comment]
@rouault I addressed all the points. could you please redo a review?


> the usual practice is to use the dummy 0 generic version:
> 
> http://www.opengis.net/def/crs/EPSG/0/4326 is a dereferencable URL, whereas http://www.opengis.net/def/crs/EPSG/9.6.2/4326 is not

@elpaso I changed everywhere the version 9.6.2 of the url to 0.

**Comment by elpaso** [comment]
> @rouault I addressed all the points. could you please redo a review?
> 
> > the usual practice is to use the dummy 0 generic version:
> > http://www.opengis.net/def/crs/EPSG/0/4326 is a dereferencable URL, whereas http://www.opengis.net/def/crs/EPSG/9.6.2/4326 is not
> 
> @elpaso I changed everywhere the version 9.6.2 of the url to 0.

Sounds good to me, it shouldn't break anything.

**Review by rouault — APPROVED** [review]

**Comment by rouault** [comment]
> Sorry to ask question after the PR was merged but does QGIS check the conformance class before trying to use crs in API requests?

no it doesn't. But it should normally only use the extension if the server advertizes a non CRS84 crs in the "crs" array member. What does your /collection response look like ?

**Comment by vincentsarago** [comment]
oh interesting, I guess it's because we return EPSG:4326 instead of CRS84 https://github.com/developmentseed/tipg/blob/main/tipg/model.py#L89

**Pull request merged**